implementation/transformer/transformer.py

```diff
@@ -0,0 +1,174 @@
+class PositionalEncoding(nn.Module):
+    def __init__(self, d_model, dropout=0.1, max_len=100):
+        super(PositionalEncoding, self).__init__()
+        self.dropout = nn.Dropout(p=dropout)
+
+        pe = torch.zeros(max_len, d_model)
+        position = torch.arange(0, max_len, dtype=torch.float).unsqueeze(1)
+        div_term = torch.exp(torch.arange(0, d_model, 2).float() * (-math.log(10000.0) / d_model))
+        pe[:, 0::2] = torch.sin(position * div_ter
+        pe[:, 1::2] = torch.cos(position * div_term)
+        pe = pe.unsqueeze(0).transpose(0, 1)
+        self.register_buffer('pe', pe)
+
+    def forward(self, x):
+        x = x + self.pe[:x.size(0), :]
+        return self.dropout(x)
+
+
+class SimpleTransformerEncoder(nn.Module):
+    """
+    Standard implementation of a Transformer Encoder.
+    Follows:
+        - Original paper diagram: https://arxiv.org/pdf/1706.03762.pdf, Figure 1
+        - Tutorials:
+            - https://pytorch.org/tutorials/beginner/transformer_tutorial.html
+            - https://andrewpeng.dev/transformer-pytorch/
+            - https://nlp.seas.harvard.edu/2018/04/03/attention.html
+    """
+    def __init__(self, vocabulary_size, dropout):
+        super().__init__()
+
+        # TODO: Do we want to use the dropout functionality of
+        # PositionalEncoder, TransformerEncoderLayer, or both?
+
+        D_MODEL = 512
+        # d_model: The following citations from
+        # https://arxiv.org/pdf/1706.03762.pdf explains what is d_model:
+        #
+        # The encoder is composed of a stack of N = 6 identical
+        # layers. Each layer has two sub-layers. The first is a multi-head
+        # self-attention mechanism, and the second is a simple, positionwise
+        # fully connected feed-forward network.
+        #
+        # The decoder is also composed of a stack of N = 6 identical layers. In
+        # addition to the two sub-layers in each encoder layer, the decoder
+        # inserts a third sub-layer, which performs multi-head attention over
+        # the output of the encoder stack
+        #
+        # All sub-layers in the model, as well as the embedding
+        # layers, produce outputs of dimension d_model = 512.
+
+        self.model_type='TransformerEncoder'
+
+        self.embedding = nn.Embedding(vocabulary_size, embedding_dimension)
+        self.positional_encoder = PositionalEncoder(
+            d_model=D_MODEL,
+            dropout,
+            max_len=self.cfg.max_ts
+        )
+
+        self.transformer_encoder = nn.modules.TransformerEncoder(
+            nn.modules.TransformerEncoderLayer(
+                d_model=D_MODEL,  # 512 used in the original paper.
+                nhead=8,  # Multi-head attention head count. 8 used in the original paper.
+                dropout=dropout,
+            ),
+            num_layers=6  # Number of sub-encoder-layers in the encoder. 6 used in the original paper.
```

Comment:
We will need much smaller dimensions, otherwise it both will overfit heavily (in original paper, they train it on MT, 4.5M pairs) and it will be computationally demanding (orig. paper - 8 P100 GPUs, 0.4s per training steps).